Add `tag` command to restic cli to manipulate tags on existing snapshots.

middelink opened the pull request:
Closes #851

Commit:
Add `tag`: Manipulate tags on existing snapshots

Add integration testing.

fd0 commented on src/cmds/restic/cmd_tag.go:
```diff
@@ -0,0 +1,189 @@
+	cmdRoot.AddCommand(cmdTag)
+
+	tagFlags := cmdTag.Flags()
+	tagFlags.StringSliceVar(&tagOptions.SetTags, "set", nil, "`tag` which will replace the existing tags, can be given multiple times")
```
Add parentheses around the `can be given multiple times`, that's the style we use for the other options, e.g. here: https://github.com/restic/restic/blob/master/src/cmds/restic/cmd_backup.go#L77

Also for the other options below, obviously.

middelink replied:
Done.

fd0 commented on src/cmds/restic/cmd_tag.go:
```diff
@@ -0,0 +1,189 @@
+		return false, nil
+	}
+
+	if len(setTags) != 0 {
```
Maybe split the code below into a function on the `Snapshot` struct, in the `restic` package?

fd0 commented on src/cmds/restic/cmd_tag.go:
```diff
@@ -0,0 +1,189 @@
+		sn.Tags = setTags
+		changed = true
+	} else {
+		for _, add := range addTags {
```
This is a candidate for a new function, `Snapshot.AddTags([]string)` or so.

fd0 commented on src/cmds/restic/cmd_tag.go:
```diff
@@ -0,0 +1,189 @@
+				changed = true
+			}
+		}
+		for _, remove := range removeTags {
```
Same here, `Snapshot.RemoveTags([]string)`

fd0 commented on src/cmds/restic/cmd_tag.go:
```diff
@@ -0,0 +1,189 @@
+
+	if !gopts.NoLock {
+		Verbosef("Create exclusive lock for repository\n")
+		lock, err := lockRepoExclusive(repo)
```
I think a non-exclusive lock is sufficient, what do you think?

middelink replied:
Hmmm. will multiple non-exclusive Locks serialize each other? Because otherwise there will be a race with say `restic snapshots`, or `restic forget` or.. well. all functions which load the list of snapshotIDs to work on.

fd0 replied:
Good point, I retract my comment :)

fd0 commented on src/cmds/restic/cmd_tag.go:
```diff
@@ -0,0 +1,189 @@
+	var ids restic.IDs
+	if len(args) != 0 {
+		// When explit snapshot-IDs are given, the filtering does not matter anymore.
+		opts.Host = ""
```
I feel it may be better to first create a list of snapshot IDs that are to be modified, either by taking the explicit list of IDs (if given) or by listing all snapshots and then applying the filters. Then iterate through that list and modify the tags. This way, `changeTags()` won't need the filter criteria at all.

middelink replied:
Wait, that is what I do, no? Build a list of snapshotIDs. The problem is, if I want to do filtering I need to Load the snapshot which, sure, I can do in `runTag()`, but then I need to lift the Loading code also into the code path for the explicit ids. And I really really would like to keep the ids.Uniq() there.

fd0 replied:
I see what you mean. Leave it ;)

fd0 commented:
Ah, and here you can drop the commits from the other PRs.

Commit:
Snapshot: Add `AddTags()` and `RemoveTags()`

Both prevent duplicate tags.

Commit:
Fix type of ID field in `cmd_snapshots` type Snapshot

Commit:
Snapshot: Add Original ID

The Original ID is used when the snapshot is modified (e.g. by `tag`
command). Adjust integration testing to assert correctness.

fd0 commented:
Ok, so the code already looks good, we can leave it this way. Two more things need to be done:
 * Add the new field (and a short paragraph which explains it) to `doc/Design.md`
 * Describe the new command in a section in `doc/Manual.md`

fd0 commented:
When you rebase to the current master branch and force-push the commits, git should drop the duplicate commits from the other PRs...

fd0 commented:
Small nitpick: Use present tense for commit descriptions and subject lines, i.e. "add foobar" instead of "added foobar"

middelink commented:
Well, the git rebase did not make things better, now we have 19 commits in the PR. My current understand of git means... hands off till it is submitted :(

fd0 commented:
This should fix it:
```
git reset --hard 3a1ecae0204bc205a0adbd710022237fdb77bd0d
git rebase master
git push -f
```

(I have a backup of your commits locally)

fd0 commented:
Exactly, when you rebase, you're rewriting history (a "non fast-forward commit" in git terms), so in that special case you need `push -f`.

Commit:
Add design and user documentation for the `restic tag` command

fd0 commented:
Hm, now you cannot use my instructions, so I'll fix it for you before merging.

middelink commented:
And... shoot. I did not see your comment after I pushed the documentation commit. Can I still run that rebase now? Or will it completely hose the world?

fd0 committed:
Documentation fixes
